lib/plugins/cpuUsageThrottle.js

```diff
@@ -0,0 +1,235 @@
+            // If we fail to stat for whatever reason, err on the side of
+            // accepting traffic, while failure probably indicates something
+            // horribly wrong with the server, identifying this error case and
+            // appropriately responding is way outside the scope of this plugin
+            if (e) {
+                // When cpu is NaN, we should never reject
+                stat = { cpu: NaN };
+            }
+
+            // Divide by 100 to match linux's top format
+            self._ewma.insert(stat.cpu / 100);
+            self._cpu = self._ewma.value();
+            // Update reject with the % of traffic we should be rejecting. This
+            // is safe since max > limit so the denominator can never be 0. If
+            // the current cpu usage is less that the limit, _reject will be
+            // negative and we will never shed load
+            self._reject =
+                (self._cpu - self._limit) / (self._max - self._limit);
+            self._timeout = setTimeout(updateReject, self._interval);
+        });
+    }
+
+    // Kick off updating our _reject value
+    updateReject();
+
+    function onRequest (req, res, next) {
+        // Check to see if this request gets rejected. Since, in updateReject,
+        // we calculate a percentage of traffic we are planning to reject, we
+        // can use Math.random() (which picks from a uniform distribution in
+        // [0,1)) to give us a `self._reject`% chance of dropping any given
+        // request. This is a stateless was to drop approximatly `self._reject`%
+        // of traffic.
+        var draw = Math.random();
+
+        if (draw >= self._reject) {
+            return next(); // Don't reject this request
+        }
+
+        var err = new self._err({
+            context: {
+                plugin: 'cpuUsageThrottle',
+                cpuUsage: self._cpu,
+                limit: self._limit,
+                max: self._max,
+                reject: self._reject,
+                halfLife: self._halfLife,
+                interval: self._interval,
+                draw: draw
+            }
+        });
+
+        return next(err);
+    }
+
+    // Allow the app to clear the timeout for this plugin if necessary, without
+    // this we would never be able to clear the event loop when letting Node
+    // shut down gracefully
+    function close () {
+        clearTimeout(self._timeout);
+    }
+    onRequest.close = close;
+
+    // Expose internal plugin state for introspection
+    onRequest.state = self;
+
+    /**
+     * cpuUsageThrottle.update
+     *
+     * Allow the plugin's configuration to be updated during runtime.
+     *
+     * @param {Object} opts The opts object for reconfiguring this plugin, it
+     *    follows the same format as the constructor for this plugin.
+     */
+    onRequest.update = function update(newOpts) {
+        assert.object(newOpts, 'newOpts');
```

Comment:
For this function you might actually want to log/emit that we've updated the parameters, to ease debugging later on.

Reply:
Don't have access to the logger here :-\

I can start caching the logger on the first request though.